src/Symfony/Contracts/HttpClient/ResponseInterface.php

```diff
@@ -0,0 +1,68 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Contracts\HttpClient;
+
+use Symfony\Contracts\HttpClient\Exception\ClientExceptionInterface;
+use Symfony\Contracts\HttpClient\Exception\ExceptionInterface;
+use Symfony\Contracts\HttpClient\Exception\RedirectionExceptionInterface;
+use Symfony\Contracts\HttpClient\Exception\ServerExceptionInterface;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+
+/**
+ * A (lazily retrieved) HTTP response.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ *
+ * @throws RedirectionExceptionInterface On a 3xx when getStatusCode() has not been called
+ * @throws ClientExceptionInterface      On a 4xx when getStatusCode() has not been called
+ * @throws ServerExceptionInterface      On a 5xx when getStatusCode() has not been called
+ * @throws TransportExceptionInterface   when a network error occurs
+ */
+interface ResponseInterface
+{
+    /**
+     * Gets the HTTP status code of the response.
+     */
+    public function getStatusCode(): int;
+
+    /**
+     * @return string[][] The headers of the response keyed by header names in lowercase
+     */
+    public function getHeaders(): array;
+
+    /**
+     * Gets the response body as a string.
+     */
+    public function getContent(): string;
+
+    /**
+     * Returns info coming from the transport layer.
+     *
+     * This method SHOULD NOT throw any ExceptionInterface and SHOULD be non-blocking.
```

Comment:
what is the behaviour if a non-blocking http request has been done and the information is not yet ready?

Reply:
The behavior is that the method should return the current state of the request/response:
e.g; `0` as http_code when it did arrive yet, `primary_ip` empty or not before/after DNS resolution, etc. That's how both curl and native clients behave also (and what the contracts imply).

Reply:
okay, that makes sense. maybe this could be said explicitly here? and the list below could explicitly state the not-yet-known values for all options, as it seems to not always be `null`. e.g. would raw_headers be an empty array or null? start_time?

Reply:
I read the comment again and it's clear to me... :)

Reply:
its nice when its clear to you. but you know what to expect. when i looked at it, it was not 100% clear to me. and there will be people reading this with less of an idea of async than me, so i think it would be better to be fully explicit. add something like:

`The list below describes what value is returned when the information is not available yet.`

(and btw, i'd say "SHOULD NOT throw any exception" - or is it fine to throw some other exception that is not the ExceptionInterface? and, what happens when requesting an info that the client does not know about / did not implement?)

Reply:
I added some related sentences. PR welcome after merge if you're up to. Thanks for asking.

Reply:
I added this:
`Implementations are not required to support all options described above; they can also support more custom options; but in any case, they MUST throw a TransportExceptionInterface when an unsupported option is passed.`